networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/RpcResponseDecoder.java

```diff
@@ -101,23 +144,41 @@ public RpcResponseDecoder(SszSchema<T> responseSchema, RpcEncoding encoding) {
 
   public void close() {
     payloadDecoder.ifPresent(ByteBufDecoder::close);
+    contextDecoder.ifPresent(ByteBufDecoder::close);
     errorDecoder.ifPresent(ByteBufDecoder::close);
   }
 
   public void complete() throws RpcException {
-    try {
-      if (payloadDecoder.isPresent()) {
-        payloadDecoder.get().complete();
-        payloadDecoder = Optional.empty();
-      }
-    } finally {
-      if (errorDecoder.isPresent()) {
-        errorDecoder.get().complete();
-        errorDecoder = Optional.empty();
-      }
+    final List<RpcException> exceptions = new ArrayList<>();
+    completeDecoder(payloadDecoder).ifPresent(exceptions::add);
+    payloadDecoder = Optional.empty();
+    completeDecoder(contextDecoder).ifPresent(exceptions::add);
+    contextDecoder = Optional.empty();
+    completeDecoder(errorDecoder).ifPresent(exceptions::add);
+    errorDecoder = Optional.empty();
+
+    if (exceptions.size() > 0) {
+      throw exceptions.get(0);
```

Comment:
We can actually add the rest of the exceptions as suppressed with `Throwable.addSuppressed` so that we get the stack trace reported if the exception winds up being unhandled/logged.

I think it can be done with:
```suggestion
      throw exceptions.stream().reduce((a, b) -> {
                a.addSuppressed(b);
                return a;
              });
```